onnxruntime/core/providers/cuda/math/topk.cc

```diff
@@ -77,19 +77,29 @@ Status TopK<inputk>::ComputeInternal(OpKernelContext* ctx) const {
   int32_t axis = static_cast<int32_t>(axis_ < 0 ? rank + axis_ : axis_);
   ORT_ENFORCE(axis > -1 && axis < rank);
 
+  int64_t k_value = 0;
   if (inputk) {
     auto tensor_K = ctx->Input<Tensor>(1);
     ORT_ENFORCE(nullptr != tensor_K);
-    K_ = *tensor_K->Data<int64_t>();
-    ORT_ENFORCE(K_ >= 0 && K_ <= tensor_X->Shape().GetDims()[axis]);
+    k_value = *tensor_K->Data<int64_t>();
+  } else {  // from attribute
+    k_value = attr_k_;
   }
 
-  auto output_shape = tensor_X->Shape();
-  output_shape[axis] = K_;
+  // Now that we know the value of 'K' and the input shape,
+  // make a final validation before going to the implementation
+  const auto& input_shape = tensor_X->Shape();
+  if ((k_value < 0) || (k_value > input_shape.GetDims()[axis])) {
```

Comment:
k=0 is also invalid.

Reply:
I think k == 0 is valid - I think if k was supposed to be >0, the spec would have specified that k be non-zero positive. In any case, if we disallow k being 0 now, it is a more restricting check. I would rather not change the check for now as we could unintentionally break something.

Reply:
It's fine that we keep current behavior.

In my understanding, zero is neither a negative nor a positive number. The spec says positive value means k > 0.
If someone accidently uses k=0, they will find it out very soon since output is empty.

Reply:
Yeah, pedantically I think you're right in that positive doesn't include zero but best not to make the check more restrictive now and in any case, as you say, we will bail out gracefully once we find that the output is going to be empty. So, not much harm.